esp-hal-common/src/gpio.rs

```diff
@@ -1485,19 +1490,38 @@ impl<MODE> embedded_hal_async::digital::Wait for AnyPin<Input<MODE>> {
 }
 
 /// General Purpose Input/Output driver
-pub struct IO {
+pub struct IO<MODE> {
     _io_mux: IO_MUX,
     pub pins: Pins,
+    _mode: PhantomData<MODE>,
 }
 
-impl IO {
-    pub fn new(gpio: GPIO, io_mux: IO_MUX) -> Self {
+#[cfg(feature = "async")]
+impl IO<AsyncMode> {
+    pub fn new_async(mut gpio: GPIO, io_mux: IO_MUX) -> IO<AsyncMode> {
+        crate::peripherals::bind_gpio_interrupt(&mut gpio, asynch::handle_gpio_interrupt);
         let pins = gpio.split();
-        let io = IO {
+        IO {
             _io_mux: io_mux,
             pins,
-        };
-        io
+            _mode: PhantomData::default(),
+        }
+    }
+}
+
+impl IO<BlockingMode> {
+    pub fn new(gpio: GPIO, io_mux: IO_MUX) -> IO<BlockingMode> {
+        let pins = gpio.split();
+        IO {
+            _io_mux: io_mux,
+            pins,
+            _mode: PhantomData::default(),
+        }
+    }
+
+    pub fn set_interrupt_handler(&mut self, handler: unsafe extern "C" fn() -> ()) {
+        let mut gpio = unsafe { crate::peripherals::GPIO::steal() };
```

Comment:
I think this probably needs to be in a critical section? maybe all the bind functions do, but especially this one as we may get a race condition updating the handler in operation 🤔